HSEARCH-2213 Better error message when time out occurs

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-2213

The log error will look something like this
```
org.hibernate.search.exception.SearchException: HSEARCH400010: Elasticsearch connection time-out; check the cluster status, it should be 'green';
 Request:
========
Operation: 
Data:
null
Response:
=========
Status: 408
Error message: 408 Request Timeout
Cluster name: elasticsearch
Cluster status: yellow
```

Sanne commented:
I'm wondering if there might be other kind of issues to cause such a TIMEOUT. i.e. what about a "traditional" stuck TCP connection?

Might be simpler to make the message slightly more generic, like:
 
    Elasticsearch connection time-out; check the cluster status.

Good idea to print out the cluster status!

DavideD commented:
I think it's be  to specify that it should be 'green' as some people might think that 'yellow' is an acceptable state. I guess I could write:
`Elasticsearch connection time-out; check the connection to the cluster and the status ( it should be green).`

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/client/impl/JestClient.java:
```diff
@@ -46,6 +48,8 @@
 
 	private static final Log LOG = LoggerFactory.make( Log.class );
 
+	private static final int TIME_OUT = 408;
```
optional: I'd probably rename this to make it clear that it's a response code?

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java:
```diff
@@ -71,4 +71,8 @@
 	void analyzerIsNotRemote(String name);
+
+	@Message(id = ES_BACKEND_MESSAGES_START_ID + 10,
+			value = "Elasticsearch connection time-out; check the cluster status, it should be 'green';\n Request:\n========\n%1$sResponse:\n=========\n%2$s" )
```
BTW some people will cry at seeing multi-line logs as they make grepping very hard.

I personally like this one though.. just saying that maybe we shouldn't :)

DavideD replied:
Yes, I'm not a big fan as well. I've been just consistent with the other messages, I think we might change them as a separate issue if someone complains

DavideD replied:
I'm also wondering if the `\n` will work on Windows

Sanne replied:
good point. Let's just avoid it then.

gunnarmorling replied:
Get another editor than Notepad and it will work ;)

Kidding aside, the request and response will contain line feeds anyways (JSON structures). Also I expect people to use tools such as Splunk, Logstash etc. when analyzing production logs. To me, it's important to get out the required context in a sensible way, so I don't think we should let stop ourselves by people stuck with their tools in the 90ies.

Sanne replied:
ok yes I have no strong opinion.. go ahead if you like it better.

Still, I'm sure that some people are stuck with their tools of the 90ths and will hate you as it might not be their choice to change that :)

DavideD commented:
* Rebased
* log message updated to
```
org.hibernate.search.exception.SearchException: HSEARCH400010: Elasticsearch connection time-out; check the connection to the cluster and the status (it should be 'green'); Request: [operation: class org.hibernate.search.backend.elasticsearch.impl.ElasticsearchIndexManager$1, data: null], Response: [error: 408 Request Timeout, response code: 408, cluster: elasticsearch, status: yellow]
```
* removed line separator char from the other error messages

Sanne commented:
great, merged!
